File: 124. Binary Tree Maximum Path Sum/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.maxSum=-1e9

    def depth(self,root) -> int:
        if root==None:
            return -1e9
        left = self.depth(root.left)
        right = self.depth(root.right)
        self.maxSum = max(self.maxSum,root.val+left+right)
        branchSum = max(root.val,max(root.val+left,root.val+right))
        self.maxSum = max(self.maxSum, branchSum)
        self.maxSum=max(self.maxSum,max(left,right))
        
        return branchSum

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if root==None:
            return -1e9
        one = self.depth(root)
        # left = self.maxPathSum(root.left)
        # right = self.maxPathSum(root.right)

        # maxSum =  max(maxSum,max(left,right))
        # maxSum = max(maxSum,one)
        return self.maxSum

        
        

File: 124. Binary Tree Maximum Path Sum/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root,int &maxSum){
        if(root==NULL){
            return -1e9;
        }
        if(root->left==NULL && root->right==NULL){
            return root->val;
        }

        int left = pathSum(root->left,maxSum);
        int right = pathSum(root->right,maxSum);
        maxSum = max(maxSum,max(root->val+left+right,max(left,right)));
        maxSum = max(max(max(left,right)+root->val,root->val),maxSum);
        return max(max(left,right)+root->val,root->val);
    }
    int maxPathSum(TreeNode* root) {
        if(root==NULL) return 0;
        if(root->left==NULL && root->right==NULL) return root->val;
        int maxm = INT_MIN;
        pathSum(root,maxm);
        return maxm;
        
    }
};